### ISSUE TITLE
Cluster join fails due to insufficient ephemeral-storage

### ISSUE THREAD

**Issue by hvenev-vmware** [kubernetes/kubernetes#99305]
#### What happened:
`kubeadm join` on the new node fails sometimes. The following error appears in the log:
```
Feb 22 14:04:33 NEWHOST kubelet[7331]: W0222 14:04:33.866713    7331 predicate.go:113] Failed to admit pod etcd-NEWHOST_kube-system(13c2d55e88e08af2d946555ab67e9eab) - Unexpected error while attempting to recover from admission failure: preemption: error finding a set of pods to preempt: no set of running pods found to reclaim resources: [(res: ephemeral-storage, q: 104857600), ]
```

#### What you expected to happen:
`kubeadm join` on the new node succeeds every time

#### How to reproduce it (as minimally and precisely as possible):
Join a control plane node

#### Anything else we need to know?:
The node has over 40 GB of free space on its root partition.

I added some more logging to the following functions:
- `pkg/kubelet/cm/container_manager_linux.go`, `Start`
- `pkg/kubelet/lifecycle/predicate.go`, `GeneralPredicates`
- `pkg/kubelet/nodestatus/setters.go`, `MachineInfo`
- `pkg/kubelet/nodestatus/setters.go`, `ReadyCondition`
- `pkg/kubelet/preemption/preemption.go`, `HandleAdmissionFailure`
- `pkg/scheduler/framework/plugins/noderesources/fit.go`, `Filter`

It appears that the ephemeral storage is detected a bit too late:
```
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.599153    7331 setters.go:284] TRACE MachineInfo start
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.599202    7331 setters.go:334] TRACE MachineInfo got initialCapacity map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.599333    7331 setters.go:408] TRACE MachineInfo done, Capacity: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.599357    7331 setters.go:409] TRACE MachineInfo done, Allocatable: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42086916096 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.666348    7331 setters.go:527] TRACE capacity cpu: {i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.666413    7331 setters.go:527] TRACE capacity memory: {i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.666425    7331 setters.go:527] TRACE capacity pods: {i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.666435    7331 setters.go:527] TRACE capacity ephemeral-storage: {i:{value:0 scale:0} d:{Dec:<nil>} s: Format:}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.673445    7331 setters.go:284] TRACE MachineInfo start
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.673490    7331 setters.go:334] TRACE MachineInfo got initialCapacity map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.673540    7331 setters.go:408] TRACE MachineInfo done, Capacity: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.673563    7331 setters.go:409] TRACE MachineInfo done, Allocatable: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42086916096 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.692186    7331 setters.go:284] TRACE MachineInfo start
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.692250    7331 setters.go:334] TRACE MachineInfo got initialCapacity map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.692306    7331 setters.go:408] TRACE MachineInfo done, Capacity: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.692325    7331 setters.go:409] TRACE MachineInfo done, Allocatable: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42086916096 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.743289    7331 setters.go:527] TRACE capacity cpu: {i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.743339    7331 setters.go:527] TRACE capacity memory: {i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.743349    7331 setters.go:527] TRACE capacity pods: {i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.743362    7331 setters.go:527] TRACE capacity ephemeral-storage: {i:{value:0 scale:0} d:{Dec:<nil>} s: Format:}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.752328    7331 container_manager_linux.go:613] TRACE containerManagerImpl.Start
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.752381    7331 container_manager_linux.go:625] TRACE containerManagerImpl.Start, map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} ephemeral-storage:{i:{value:51384074240 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.772730    7331 setters.go:527] TRACE capacity cpu: {i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.772773    7331 setters.go:527] TRACE capacity memory: {i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.772780    7331 setters.go:527] TRACE capacity pods: {i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.772786    7331 setters.go:527] TRACE capacity ephemeral-storage: {i:{value:0 scale:0} d:{Dec:<nil>} s: Format:}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.826100    7331 setters.go:284] TRACE MachineInfo start
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.826146    7331 setters.go:334] TRACE MachineInfo got initialCapacity map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} ephemeral-storage:{i:{value:51384074240 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.826353    7331 setters.go:408] TRACE MachineInfo done, Capacity: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} ephemeral-storage:{i:{value:51384074240 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.826414    7331 setters.go:409] TRACE MachineInfo done, Allocatable: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} ephemeral-storage:{i:{value:46245666740 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42086916096 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.846712    7331 predicate.go:225] TRACE nodeInfo.Allocatable &{MilliCPU:12000 Memory:42086916096 EphemeralStorage:0 AllowedPodNumber:210 ScalarResources:map[hugepages-1Gi:0 hugepages-2Mi:0]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.846749    7331 predicate.go:226] TRACE nodeInfo.Requested &{MilliCPU:0 Memory:0 EphemeralStorage:0 AllowedPodNumber:0 ScalarResources:map[]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.846845    7331 predicate.go:225] TRACE nodeInfo.Allocatable &{MilliCPU:12000 Memory:42086916096 EphemeralStorage:0 AllowedPodNumber:210 ScalarResources:map[hugepages-1Gi:0 hugepages-2Mi:0]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.846854    7331 predicate.go:226] TRACE nodeInfo.Requested &{MilliCPU:0 Memory:0 EphemeralStorage:0 AllowedPodNumber:0 ScalarResources:map[]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.846892    7331 predicate.go:225] TRACE nodeInfo.Allocatable &{MilliCPU:12000 Memory:42086916096 EphemeralStorage:0 AllowedPodNumber:210 ScalarResources:map[hugepages-1Gi:0 hugepages-2Mi:0]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.846940    7331 predicate.go:226] TRACE nodeInfo.Requested &{MilliCPU:0 Memory:0 EphemeralStorage:0 AllowedPodNumber:0 ScalarResources:map[]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.847005    7331 predicate.go:225] TRACE nodeInfo.Allocatable &{MilliCPU:12000 Memory:42086916096 EphemeralStorage:0 AllowedPodNumber:210 ScalarResources:map[hugepages-1Gi:0 hugepages-2Mi:0]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.847014    7331 predicate.go:226] TRACE nodeInfo.Requested &{MilliCPU:0 Memory:0 EphemeralStorage:0 AllowedPodNumber:0 ScalarResources:map[]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.847072    7331 predicate.go:225] TRACE nodeInfo.Allocatable &{MilliCPU:12000 Memory:42086916096 EphemeralStorage:0 AllowedPodNumber:210 ScalarResources:map[hugepages-1Gi:0 hugepages-2Mi:0]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.847080    7331 predicate.go:226] TRACE nodeInfo.Requested &{MilliCPU:250 Memory:0 EphemeralStorage:0 AllowedPodNumber:0 ScalarResources:map[]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.847160    7331 predicate.go:225] TRACE nodeInfo.Allocatable &{MilliCPU:12000 Memory:42086916096 EphemeralStorage:0 AllowedPodNumber:210 ScalarResources:map[hugepages-1Gi:0 hugepages-2Mi:0]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.847177    7331 predicate.go:226] TRACE nodeInfo.Requested &{MilliCPU:450 Memory:0 EphemeralStorage:0 AllowedPodNumber:0 ScalarResources:map[]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.866562    7331 predicate.go:225] TRACE nodeInfo.Allocatable &{MilliCPU:12000 Memory:42086916096 EphemeralStorage:0 AllowedPodNumber:210 ScalarResources:map[hugepages-1Gi:0 hugepages-2Mi:0]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.866598    7331 predicate.go:226] TRACE nodeInfo.Requested &{MilliCPU:550 Memory:0 EphemeralStorage:0 AllowedPodNumber:0 ScalarResources:map[]}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.866620    7331 preemption.go:67] TRACE cannot admit etcd-NEWHOST, [Node didn't have enough resource: ephemeral-storage, requested: 104857600, used: 0, capacity: 0]
Feb 22 14:04:33 NEWHOST kubelet[7331]: W0222 14:04:33.866713    7331 predicate.go:113] Failed to admit pod etcd-NEWHOST_kube-system(13c2d55e88e08af2d946555ab67e9eab) - Unexpected error while attempting to recover from admission failure: preemption: error finding a set of pods to preempt: no set of running pods found to reclaim resources: [(res: ephemeral-storage, q: 104857600), ]
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.905188    7331 setters.go:527] TRACE capacity cpu: {i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.905231    7331 setters.go:527] TRACE capacity memory: {i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.905239    7331 setters.go:527] TRACE capacity pods: {i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}
Feb 22 14:04:33 NEWHOST kubelet[7331]: E0222 14:04:33.905248    7331 setters.go:527] TRACE capacity ephemeral-storage: {i:{value:51384074240 scale:0} d:{Dec:<nil>} s: Format:BinarySI}
Feb 22 14:04:43 NEWHOST kubelet[7331]: E0222 14:04:43.789009    7331 setters.go:284] TRACE MachineInfo start
Feb 22 14:04:43 NEWHOST kubelet[7331]: E0222 14:04:43.789081    7331 setters.go:334] TRACE MachineInfo setting capacity to map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} ephemeral-storage:{i:{value:51384074240 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI}]
Feb 22 14:04:43 NEWHOST kubelet[7331]: E0222 14:04:43.789183    7331 setters.go:408] TRACE MachineInfo done, Capacity: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} ephemeral-storage:{i:{value:51384074240 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42191773696 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
Feb 22 14:04:43 NEWHOST kubelet[7331]: E0222 14:04:43.789234    7331 setters.go:409] TRACE MachineInfo done, Allocatable: map[cpu:{i:{value:12000 scale:-3} d:{Dec:<nil>} s: Format:DecimalSI} ephemeral-storage:{i:{value:46245666740 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-1Gi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} hugepages-2Mi:{i:{value:0 scale:0} d:{Dec:<nil>} s: Format:BinarySI} memory:{i:{value:42086916096 scale:0} d:{Dec:<nil>} s: Format:BinarySI} pods:{i:{value:210 scale:0} d:{Dec:<nil>} s: Format:DecimalSI}]
```

/sig node storage

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.4, commit e87da0bd6e03ec3fea7933c4b5263d151aafd07c
- Kernel (e.g. `uname -a`): 4.9.252-1.ph2

**Comment by pacoxu** [comment]
/assign 
I will try to reproduce it today.

**Comment by pacoxu** [comment]
LocalStorageCapacityIsolation is a feature gate for kubelet. Can you check if you disable it?

https://github.com/kubernetes/kubernetes/blob/607423871020779daa6a106bf2ee91969789790e/pkg/kubelet/nodestatus/setters.go#L517-L520

ReadyCondition will check ephemeral-storage as required capacities if this feature gate is enabled in kubelet configuration.

Would you check it first?

**Comment by pacoxu** [comment]
- [x] 1.8 alpha  ephemeral-storage
- [x] 1.10 Promote LocalStorageCapacityIsolation feature to beta #60159
- [x] 1.15 KEPs Quotas for ephemeral-storage for https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1029-ephemeral-storage-quotas#quotas-for-ephemeral-storage LocalStorageCapacityIsolationFSQuotaMonitoring default to false is Alpha since 1.15
- [ ] Failed to promote LocalStorageCapacityIsolationFSQuotaMonitoring feature to beta #79041 
- [x] 1.20 Honor disabled LocalStorageCapacityIsolation in scheduling #96092 

I am no sure whether there is a plan to GA LocalStorageCapacityIsolation?

**Comment by pacoxu** [comment]
I tried to fix it in #99635.
Need some testing later.

**Comment by hvenev-vmware** [comment]
> LocalStorageCapacityIsolation is a feature gate for kubelet. Can you check if you disable it?
> 
> https://github.com/kubernetes/kubernetes/blob/607423871020779daa6a106bf2ee91969789790e/pkg/kubelet/nodestatus/setters.go#L517-L520
> 
> ReadyCondition will check ephemeral-storage as required capacities if this feature gate is enabled in kubelet configuration.
> 
> Would you check it first?

It appears that the pod fails before `ReadyCondition` sees the storage capacity:

```
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.146292     468 setters.go:338] MachineInfo initialCapacity[ephemeral-storage] not found
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.146326     468 setters.go:389] MachineInfo finished setting node.Status.Allocatable
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.251632     468 setters.go:531] ReadyCondition requiredCapacities=[cpu memory pods ephemeral-storage]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.251674     468 setters.go:532] ReadyCondition missingCapacities=[ephemeral-storage]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.263142     468 setters.go:338] MachineInfo initialCapacity[ephemeral-storage] not found
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.263192     468 setters.go:389] MachineInfo finished setting node.Status.Allocatable
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.266659     468 setters.go:338] MachineInfo initialCapacity[ephemeral-storage] not found
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.266718     468 setters.go:389] MachineInfo finished setting node.Status.Allocatable
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.355633     468 setters.go:531] ReadyCondition requiredCapacities=[cpu memory pods ephemeral-storage]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.356329     468 setters.go:532] ReadyCondition missingCapacities=[ephemeral-storage]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.370696     468 setters.go:531] ReadyCondition requiredCapacities=[cpu memory pods ephemeral-storage]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.370737     468 setters.go:532] ReadyCondition missingCapacities=[ephemeral-storage]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.394507     468 fit.go:230] Fits nodeInfo.Allocatable.EphemeralStorage=0
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.394615     468 fit.go:230] Fits nodeInfo.Allocatable.EphemeralStorage=0
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.394660     468 fit.go:230] Fits nodeInfo.Allocatable.EphemeralStorage=0
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.394732     468 fit.go:230] Fits nodeInfo.Allocatable.EphemeralStorage=0
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.394812     468 fit.go:230] Fits nodeInfo.Allocatable.EphemeralStorage=0
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.394892     468 fit.go:230] Fits nodeInfo.Allocatable.EphemeralStorage=0
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.395057     468 fit.go:230] Fits nodeInfo.Allocatable.EphemeralStorage=0
Mar 02 14:48:22 NEWHOST kubelet[468]: W0302 14:48:22.395141     468 predicate.go:113] Failed to admit pod etcd-NEWHOST_kube-system(7446b9f3480fa54d0333f6e32bb4c224) - Unexpected error while attempting to recover from admission failure: preemption: error finding a set of pods to preempt: no set of running pods found to reclaim resources: [(res: ephemeral-storage, q: 104857600), ]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.402260     468 setters.go:335] MachineInfo initialCapacity[ephemeral-storage]={i:{value:51384074240 scale:0} d:{Dec:<nil>} s: Format:BinarySI}
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.402399     468 setters.go:389] MachineInfo finished setting node.Status.Allocatable
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.486794     468 setters.go:531] ReadyCondition requiredCapacities=[cpu memory pods ephemeral-storage]
Mar 02 14:48:22 NEWHOST kubelet[468]: E0302 14:48:22.486839     468 setters.go:532] ReadyCondition missingCapacities=[]
```

**Comment by pacoxu** [comment]
Would you try this  https://github.com/pacoxu/kubernetes/releases/download/1.21.0-beta-dev/kubelet ?

I build it with `bash build/run.sh make kubelet` on #99635 PR branch.

**Comment by hvenev-vmware** [comment]
I tried to apply the PR to 1.20.4. After four successful `kubeadm join`s, one failed. I got the following error:

```
Mar 05 07:15:39 NEWHOST kubelet[8106]: I0305 07:15:39.477904    8106 container_manager_linux.go:262] failed to get rootfs info during initial: unable to find data in memory cache
```

**Comment by r2dedios** [comment]
Hi,

Same problem here. I'm using K8s v1.20.4 with CentOS7 (2 cores, 4GB and 20GB storage)

Cluster creation with _kubeadm init_ always works, but when I try to add several master nodes with _kubeadm join_ it takes a lot of time waiting for **etcd** pod to be created. In most of cases, **etcd** pod never runs. The rest of components (APIserver, controller, scheduler and proxy) exists and works (except APIServer which reboots trying to connect the missing etcd)

Checking the kubelet logs in the missing etcd hosts I have found this:

`
Mar 10 14:23:27 kmaster-02 kubelet[9448]: W0310 14:23:27.586165    9448 predicate.go:113] Failed to admit pod etcd-kmaster-02_kube-system(a9bd496345f6709b747bfbcf4a8cac0d) - Unexpected error while attempting to recover from admission failure: preemption: error finding a set of pods to preempt: no set of running pods found to reclaim resources: [(res: ephemeral-storage, q: 104857600), ]                 
`

That node have 17GB storage free. 


If kubelet is restarted, the etcd (pod and container in docker ps -a) appears.

**Comment by mindlift** [comment]
Thanks for the workaround @heilerich. I'm encountering this same issue and the workaround worked for me.

**Comment by pacoxu** [comment]
I updated #99635 
Add a start-delay of 10 seconds for a condition that the capacity of node ephemeral-storage is 0.

Not sure if this will fix it.

**Comment by neolit123** [comment]
we are starting to see reports about this every week.
@kubernetes/sig-node-bugs @kubernetes/sig-storage-bugs

**Comment by dims** [comment]
cc @adisky

**Comment by jingxu97** [comment]
cc @jingxu97

**Comment by msau42** [comment]
From https://github.com/kubernetes/kubernetes/issues/99305#issuecomment-802805653 it sounds like the issue is kubeadm etcd ephemeral-storage requests are too big for some environments?

So far this is sounding like a kubeadm issue, not a kubernetes issue.

**Comment by neolit123** [comment]
@msau42 

it seems to be something else. i doubt that users are trying machines that don't satisfy 100Mi.
for example, the report in the OP points out that the machine has 40GB.

**Comment by mindlift** [comment]
I can confirm that the VMs on which I installed k8s control plane components had sufficient disk space when I encountered the problem. I'm running Kubernetes v1.20.4 and kubeadm v1.20.4.

**Comment by jingxu97** [comment]
I think the root cause here https://github.com/kubernetes/kubernetes/issues/99305#issuecomment-794926298 sounds right. There is kind of race here. #99635 adding delay will help, but it does not have 100% guarantee. Not sure whether it is easy to add logic to make sure capacity is reported correctly before starting pods..

**Comment by msau42** [comment]
Did something change in 1.20 regarding the ordering of cadvisor and node startup?

**Comment by pacoxu** [comment]
@jingxu97  
If the reason would be a disk problem(or something we can figure out) and the delay can solve it, I think we should add a warning besides the delay.

**Comment by neolit123** [comment]
We need to figure out what triggers the race condition. This only happens on some machines. I have not seen this myself.

**Comment by CecileRobertMichon** [comment]
We're able to repro this nearly 100% of the time with Cluster API Provider Azure + k8s 1.20.x + Ubuntu 18.04 + default template, which makes k8s 1.20 pretty much unusable at this point. Trying to see if this repros with other non-Azure cluster-api providers with kubeadm 1.20.

**Comment by neolit123** [comment]
i can see this failing for any static pod with ephemeral storage.
that is, unless there is proof that this is somehow specific to the kubeadm static pod for etcd that kubeadm deploys.

**Comment by jackfrancis** [comment]
@neolit123 I agree that the specific kubeadm-delivered etcd pod is not interesting. (In fact it's very permissive, it's only asking for 100Mi.)

What definitely seems to be happening (in Azure) is that kubelet takes a long time to know about its underlying node's ephemeral-storage availability.

Kubelet starts:

```
Apr 22 18:56:53 acse-test-36749-control-plane-8wv76 systemd[1]: Started kubelet: The Kubernetes Node Agent.
```

~2 mins later, it complains that it can't schedule etcd:

```
Apr 22 18:58:45 acse-test-36749-control-plane-8wv76 kubelet[2598]: W0422 18:58:45.532198    2598 predicate.go:113] Failed to admit pod etcd-acse-test-36749-control-plane-8wv76_kube-system(2e306269d35643ba78288d49aad5a714) - Unexpected error while attempting to recover from admission failure: preemption: error finding a set of pods to preempt: no set of running pods found to reclaim resources: [(res: ephemeral-storage, q: 104857600), ]
```

Again, why does this not repro on nodes created by `kubeadm init`?

**Comment by neolit123** [comment]
it's also seems to be an even more scoped problems because it's not reproducible for me locally (i create a number of clusters every day, usually), the kubeadm CI or in other high level projects that use kubeadm (e.g. kind, minikube, etc).

**Comment by DennisGlindhart** [comment]
Experiencing same problem on openSuSE Kubic/Tumbletweed snapshot 20210423 with kubeadm/kubelet 1.20.2, kernel 5.11.15 running on Xen, 2 CPU, 4GB ram 30GB disk (BRTFS)

**Comment by DennisGlindhart** [comment]
> Experiencing same problem on openSuSE Kubic/Tumbletweed snapshot 20210423 with kubeadm/kubelet 1.20.2, kernel 5.11.15 running on Xen, 2 CPU, 4GB ram 30GB disk (BRTFS)

Problem consistently appear/disappear when switching between snapshots 20210401 (Working) and 20210423, so a package update in that period has probably triggered this.

Following files contains changes between those two snapshots.

http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210406.txt
http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210408.txt
http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210414.txt
http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210415.txt
http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210418.txt
http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210420.txt
http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210422.txt
http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210423.txt

**Comment by DennisGlindhart** [comment]
A bit of luck combined with 6th sense i got it narrowed down to systemd update here http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210408.txt

Looking at the changelog, one item sticks out:

```
- Default to the "unified" cgroup hierarchy. At this point, most
  users of cgroup (such as docker, libvirt, kubernetes) should be
  ready for this change. It's still possible to switch back to the
  old "hybrid" hierarchy by passing "systemd.unified_cgroup_hierarchy=0"
  option to the kernel command line.
```

Anyone else enabled cgroupsv2 on i.e. the CentOS cases (it needs to be enabled explicitly for that AFAIK)?

**Comment by jackfrancis** [comment]
@DennisGlindhart ACK on your latest respond regarding unified cgroup hierarchy.

Is there a quick, cross-distro way to verify that any folks who are repro'ing have a similar systemd config?

**Comment by DennisGlindhart** [comment]
> Is there a quick, cross-distro way to verify that any folks who are repro'ing have a similar systemd config?

The precence of file/dir /sys/fs/cgroup/cgroup.controllers should mean you are using cgroupsv2 ( Ref: https://github.com/opencontainers/runc/blob/master/docs/cgroup-v2.md#am-i-using-cgroup-v2 )

I can confirm that the file is present on the kubic snapshot which is **not** working. It is not present on the one that **does** work.

( I'm using cri-o as runtime which should support cgroupsv2 )

cgroup2 support was introduced for K8S in 1.19 ( https://github.com/kubernetes/kubernetes/pull/85218 ).

**Comment by jackfrancis** [comment]
> > Is there a quick, cross-distro way to verify that any folks who are repro'ing have a similar systemd config?
> 
> The precence of file/dir /sys/fs/cgroup/cgroup.controllers should mean you are using cgroupsv2 ( Ref: https://github.com/opencontainers/runc/blob/master/docs/cgroup-v2.md#am-i-using-cgroup-v2 )
> 
> I can confirm that the file is present on the kubic snapshot which is **not** working. It is not present on the one that **does** work.
> 
> ( I'm using cri-o as runtime which should support cgroupsv2 )
> 
> cgroup2 support was introduced for K8S in 1.19 ( #85218 ).

Thanks. So tl;dr your working theory is that cgroupsv2 + kubelet ephemeral-storage admission has some pathologies. I'll post quickly with what my repro containerd + Ubuntu 18.04-LTS OS surface area looks like.

**Comment by jasons42** [comment]
I ran into this issue consistently on Ubuntu 20.04.2 LTS using cri-o but `/sys/fs/cgroup/cgroup.controllers` is not present on my systems.

**Comment by jackfrancis** [comment]
I can also confirm that cgroupsv2 is not a factor in my repro'ing environment:

```
capi@acse-test-capz-42b7f-control-plane-svpxc:~$ uname -a
Linux acse-test-capz-42b7f-control-plane-svpxc 5.4.0-1044-azure #46~18.04.1-Ubuntu SMP Fri Mar 26 23:39:16 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
capi@acse-test-capz-42b7f-control-plane-svpxc:~$ sudo test -f /sys/fs/cgroup/cgroup.controllers
capi@acse-test-capz-42b7f-control-plane-svpxc:~$ echo $?
1
capi@acse-test-capz-42b7f-control-plane-svpxc:~$ sudo test -d /sys/fs/cgroup/cgroup.controllers
capi@acse-test-capz-42b7f-control-plane-svpxc:~$ echo $?
1
```

**Comment by jackfrancis** [comment]
Also, I'm not using cri-o, containerd here:

```
capi@acse-test-capz-42b7f-control-plane-svpxc:~$ sudo crictl version
Version:  0.1.0
RuntimeName:  containerd
RuntimeVersion:  v1.4.4
RuntimeApiVersion:  v1alpha2
```

**Comment by DennisGlindhart** [comment]
Ok - so much for that theory.

But still: What I did to narrow it to systemd-update was taking the working kubic snapshot (20210401) and run *zypper update systemd* which updates systemd (along with systemd-sysvinit and udev packages to same version). That consistently made it fail.

So I think a pointer to the root-cause might still be hiding somewhere around this changelog - even if it is not cgroupsv2. That part was just a theory (seemed like a "big" thing), but the update of systemd package is fairly hard evidence.

http://mirrors.dotsrc.org/opensuse/tumbleweed/iso/Changes.Kubic.20210408.txt
```
==== systemd ====
Version update (246.11 -> 246.13)
Subpackages: libsystemd0 libudev1 systemd-logger systemd-sysvinit udev

- Import commit 14581e01203df7aa63c7c8383a12e6ebe258476f (merge of v246.13)
  423b1e759c Revert "resolved: gracefully handle with packets with too large RR count" (bsc#1183745)
  4723778738 meson.build: make xinitrcdir configurable (bsc#1183408)
  [...]
  For a complete list of changes, visit:
  https://github.com/openSUSE/systemd/compare/9753d1c17545a5d46530696cb14254f5f12024f1...14581e01203df7aa63c7c8383a12e6ebe258476f
- Drop 0001-Revert-resolved-gracefully-handle-with-packets-with-.patch
  as it's part of v246.13.
- Make use of the new build option to ship xinitrc in
  /usr/etc/X11/xinit/xinitrc.d (bsc#1183408)
- Add 0001-Revert-resolved-gracefully-handle-with-packets-with-.patch
  Temporary workaround for bsc#1183745 (upstream issue 18917) until an
  actual fix is found.
- Default to the "unified" cgroup hierarchy. At this point, most
  users of cgroup (such as docker, libvirt, kubernetes) should be
  ready for this change. It's still possible to switch back to the
  old "hybrid" hierarchy by passing "systemd.unified_cgroup_hierarchy=0"
  option to the kernel command line.
```

**Comment by jackfrancis** [comment]
I have identified where these particular symptoms (probably) originate, and cursory testing suggests something like this eliminates the issue:

https://github.com/kubernetes/kubernetes/pull/101710/files#diff-1046a664dcd001985cf5a688c28959d4bafb964343b0d9faf3ed88e1a14c801bR69

I'll be advocating for a few mins in today's sig-node discussion if any folks following this thread want to chime in.

cc @pacoxu

**Comment by jackfrancis** [comment]
I've also observed that on some occasions the node allocatable storage is in a continual state of reporting "0". Restarting kubelet immediately fixed that.

More debugging to do....

**Comment by neolit123** [comment]
@jackfrancis did you manage to reproduce this with an arbitrary static pod that is not the etcd one?

**Comment by jackfrancis** [comment]
@neolit123 I was not able to reproduce in a non-kubeadm test, but I suspect that the way kubelet bootstrap is implemented there (and possibly environmental reasons) the static pod handling was always happening after this code executed:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/container_manager_linux.go#L640

(the above is in the containerManager `Start()` method)

kubeadm w/ ephemeral-storage requirements in etcd is the most effective repro at this time

**Comment by neolit123** [comment]
the writing of the etcd pod manifest in `/etc/kubernetes/manifests` during `kubeadm join` is [delayed](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go#L140-L152) after the kubelet has already started and i wonder if that is contributing to the observed failures. probably not given this is a cadvisor related problem. to verify that i would write an arbitrary static pod with the same request in `/etc/kubernetes/manifests`:
- before kubeadm join (which also starts the kubelet)
- right after the kubelet has started

if `LocalStorageCapacityIsolation` is causing such racy problems i don't think it should be graduated to GA.
though, it has been Beta for ~12 releases.

**Comment by jackfrancis** [comment]
Based on your assessment @neolit123, do you think we should remove the ephemeral-storage requirement from default kubeadm etcd static pod? I can make that change.

**Comment by neolit123** [comment]
based on the investigations here (thanks all), it feels like `ephemeral_storage` is a feature that may or may not work (for static pods at least).

it doesn't feel right to mitigate a kubelet bug on the side of the deployer, but given how many reports there are about this i think it should be done at this point.

happy to LGTM approve your PRs.
note it must also be backported to all versions in support (back to 1.19).

suggesting the following release note:
```
kubeadm: remove the "ephemeral_storage" request from the etcd static Pod that kubeadm deploys on stacked etcd control plane nodes. This request has caused sporadic failures on some setups due to a problem in the kubelet with cadvizor and the LocalStorageCapacityIsolation feature gate. See this issue for more details: https://github.com/kubernetes/kubernetes/issues/99305
```

we should keep this issue open, too.

**Comment by jackfrancis** [comment]
Sounds good. I'll submit a PR soon with the kubeadm change, and I'll continue to drive toward a root cause solution on the kubelet side. I'll bring this up at sig-node tomorrow.

Thanks!

**Comment by brsolomon-deloitte** [comment]
https://github.com/kubernetes/kubernetes/issues/102715 is a possible duplicate of this issue - seeing a CrashLoopBackOff failure for kube-apiserver under a similar scenario to those mentioned here, e.g. failure to join a third control plane node with mentions of `ephemeral-storage` in the kubelet logs.

**Comment by Frankkkkk** [comment]
An ugly as sin workaroud:
```
shell: "(sleep 30s && systemctl restart kubelet) & {{cluster_join_command}}"
```
(please don't hate me :D )

**Comment by brsolomon-deloitte** [comment]
It seems like this could be closed with https://github.com/kubernetes/kubernetes/pull/102673 and related cherry picks now being part of the newest releases?

Bumping to 1.21.2 allowed me to remove the `--experimental-patches` hack for etcd and several test deployments have steered clear entirely of the original issue. Thanks @jackfrancis

**Comment by jackfrancis** [comment]
@brsolomon-deloitte great news.

I'd prefer to pin this issue to this PR:

https://github.com/kubernetes/kubernetes/pull/101882

That PR will deliver more race condition resilience for *all* static pod + ephemeral-storage requirement scenarios.
